### PR TITLE
feat(cli): support dynamic imports from cjs

### LINKED ISSUE #8242: CLI: Support dynamic imports in backend code
## Feature Suggestion

We should add support for `import()` in backend code, that is dynamic `import()` expressions that do **not** got transpiled into a `require()` call but stay as is. By supporting dynamic imports, we'll make it possible to import ESM modules into Node.js CommonJS modules, and move forward with overall migration towards using the new module syntax in Node.js.

Dynamic imports have been supported in Node.js LTS since 14, so it is safe to pass them through in our build process as we currently only support Node.js 14 and 16. Note that we do not want to start transpiling all our Node.js code to ESM yet, as this would break existing usage of our packages. We rather want to add support for `import()`s in our CommonJS build process.

## Possible Implementation

We'll need to update both the build and bundle transpilation in the CLI. The building, which is what creates a production build of a single package, is handled by rollup. The configuration for that lives [here](https://github.com/backstage/backstage/blob/ca9176903623879d6c6cd41c6e5ea0c6ab083464/packages/cli/src/lib/builder/config.ts#L66), and we'll want to update our CommonJS output configuration to pass through dynamic imports.

The bundling, which we primarily used for our backend development setup, is done by webpack, and reuses a lot of our frontend webpack config. The transpilation setup for that lives [here](https://github.com/backstage/backstage/blob/ca9176903623879d6c6cd41c6e5ea0c6ab083464/packages/cli/src/lib/bundler/transforms.ts#L52), but to be honest I don't know whether this already transforms things like we want to or not, I imagine it actually already does. The last piece would be the webpack backend config itself, which lives [here](https://github.com/backstage/backstage/blob/ca9176903623879d6c6cd41c6e5ea0c6ab083464/packages/cli/src/lib/bundler/config.ts#L232), and again I'm not actually sure whether this needs updating or not, but I would imagine it does because of the `target: 'node'`

### PR DESCRIPTION
**Pull request by sennyeya**
## Hey, I just made a Pull Request!

might fix #8242? 

Context: I ran into the ESM module support issue while testing out `ink` + a new Backstage CLI.  It would be great to start the ball rolling on adding support and this seems like the first blocker.

This PR adds a new plugin that handles rollup's `resolveDynamicPlugin` and `renderDynamicPlugin` methods. Currently, I enabled it just for non-relative files as the bundler doesn't pick up relative `.mjs` files directly. Given the context of where this would be used, I doubt that that is a necessary usecase as well. I did see that the commonjs plugin is set up to run against just node_modules, 
```
       commonjs({
          include: /node_modules/,
          exclude: [/\/[^/]+\.(?:stories|test)\.[^/]+$/],
        }),
```
Not sure if that impacts this work. From what I can tell, the resolutions for the dynamic imports work despite this, but it may require further investigation. Interestingly, I only needed to change the build config, and was able to leave the webpack config untouched.

Testing plan: I'm not quite sure what else should be tested here. I ran both the `build` and `start` commands with dynamic imports and they worked as expected (like the snapshot in the rollup test). Would love ideas here.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by sennyeya** [comment]
The E2E test failure is definitely related to this work,
> [yarn new --select plugin --option id=test].err: TypeError: actionFunc is not a function

I'll investigate this week.

File: packages/cli/src/commands/index.ts [hunk 1]
```diff
@@ -437,6 +437,7 @@ function lazy(
   return async (...args: any[]) => {
     try {
       const actionFunc = await getActionFunc();
+      console.log(actionFunc, getActionFunc);
```
Comment (Rugvip) [inline packages/cli/src/commands/index.ts:440]: Nop :p

File: packages/cli/src/lib/builder/DynamicImportPreservePlugin.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Plugin } from 'rollup';
+
+export default function dynamicImportPreservePlugin(): Plugin {
```
Comment (Rugvip) [inline packages/cli/src/lib/builder/DynamicImportPreservePlugin.ts:18]: I was surprised there was nothing available for this wheen looking into it a while ago. Is there still really nothing on NPM or new config in rollup that lets you do this without a custom plugin? :sweat_smile:

**Review by Rugvip — COMMENTED** [review]
Awesome! :tada: :clap: 

Big step towards module support. Gonna play it safe and ship this after the next main-line.

File: packages/cli/src/commands/index.ts (hunk 1 above)
Comment (sennyeya) [reply packages/cli/src/commands/index.ts:440]: Whoops 😅

File: packages/cli/src/lib/builder/DynamicImportPreservePlugin.ts (hunk 2 above)
Comment (sennyeya) [reply packages/cli/src/lib/builder/DynamicImportPreservePlugin.ts:18]: I really wasn't able to find anything. This came from a thread where someone wanted to do something similar but it wasn't a full plugin.

**Comment by freben** [comment]
@Rugvip this was still of interest right

**Comment by Rugvip** [comment]
@freben yep 👍

**Comment by aramissennyeydd** [comment]
We should test this with the new backend before merging, there might be issues with the default exports there? I tested this with the CLIs and ended up moving away from default imports for those.

**Comment by aramissennyeydd** [comment]
Verified that this doesn't affect the new backend.

File: packages/cli/src/lib/builder/DynamicImportPreservePlugin.ts (hunk 2 above)
Comment (aramissennyeydd) [reply packages/cli/src/lib/builder/DynamicImportPreservePlugin.ts:18]: Turns out this is actually supported with a rollup option, https://rollupjs.org/configuration-options/#output-dynamicimportincjs. Adjusting this to use that property.

Comment (Rugvip) [reply packages/cli/src/lib/builder/DynamicImportPreservePlugin.ts]: Nice! 👍, thought something like this would be added to rollup eventually 😅

**Comment by Rugvip** [comment]
@aramissennyeydd 👍 

> only dynamically import external files

Could you elaborate a bit on this? do you mean that `import(...)` would only get forwarded if the target is relative? We'd really want it to be forwarded regardless

**Comment by aramissennyeydd** [comment]
@Rugvip (sorry for the late response here) I think I misread the webpack option and assumed it was doing both imports. When I was initially working on this, I noticed some fairly odd behavior around relative dynamic imports with rollup. I just pushed a commit with such a broken state in `plugins/catalog-backend` and `packages/backend-next`. The output of `plugins/catalog-backend` doesn't respect the dynamic import and in fact converts my MJS to CJS, `packages/backend-next` explodes with a `ERR_UNSUPPORTED_DIR_IMPORT` error. I'm pretty squarely out of my depth here 🙃, so not too sure what good next steps are here. Was thinking of opening issues outlining what I was running into in both projects.

**Comment by Rugvip** [comment]
@aramissennyeydd alright, thank you for the update 👍 

Could let it sit for the time being if you want, I might poke around at this some more at some point too.

Regarding Webpack we'll get rid of that in not too long for backend builds, it'll be part of deprecating support for the old backend system.

**Comment by aramissennyeydd** [comment]
Ah, sorry, had meant rollup 😅 . I'll try and wade through the rollup/webpack/swc miasma next week and see if I can come up with anything. The mismatch between documented options and output is bugging me a little.

**Comment by aramissennyeydd** [comment]
@Rugvip I think I'm getting closer, the SWC maintainer pointed me to an `ignoreDynamic` option that keeps the imports as is - which breaks some functionality on our end 😅 . Namely the CLI importing pattern and allowing relative dynamic imports in the backend definition,
```
lazy(() => import('./start').then(m=>m.command))
```
and 
```
backend.add(import('./authGithubProvider'))
```
Those were both being compiled to require statements with the correct file mappings (not necessarily the same as the source files). I don't think we can have it both ways of using imports and expecting some compiled and some uncompiled code.

It also means we'll have to adjust package definitions to better support ESM resolution, TS is not well behaved here (https://github.com/TypeStrong/ts-node/issues/1997). The error on our side looks like,
```
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/aramis.sennyey/Projects/backstage/backstage/plugins/catalog-backend-module-scaffolder-entity-model/src/index.ts
```

I did to a few PRs here, basically adjusting the `main` field to point to a `dist/file.cjs.js` file instead of the TS allows us to use dynamic imports. Unfortunately, that would end up being a breaking change as you couldn't use any plugins with dynamic import that don't have this correctly defined. Open to ideas here, I know there's a similar in between step of `publishConfig` which points to these files as you would expect.

**Pull request closed without merging**